src/analysis/retail/evoker/preservation/modules/talents/Echo.tsx

```diff
@@ -0,0 +1,175 @@
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, { HealEvent } from 'parser/core/Events';
+import DonutChart from 'parser/ui/DonutChart';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+import { ECHO_HEALS, SPELL_COLORS } from '../../constants';
+import { isFromHardcastEcho, isFromTAEcho } from '../../normalizers/CastLinkNormalizer';
+import HotTrackerPrevoker from '../core/HotTrackerPrevoker';
+
+class Echo extends Analyzer {
+  static dependencies = {
+    hotTracker: HotTrackerPrevoker,
+  };
+
+  protected hotTracker!: HotTrackerPrevoker;
+
+  // Map<spellId, totalHealing>, only update for echo healing
+  echoHealingBySpell: Map<number, number> = new Map<number, number>();
+  taEchoHealingBySpell: Map<number, number> = new Map<number, number>();
+
+  constructor(options: Options) {
+    super(options);
+    ECHO_HEALS.forEach((spell) => {
+      this.echoHealingBySpell.set(spell.id, 0);
+      this.taEchoHealingBySpell.set(spell.id, 0);
+    });
+    this.addEventListener(Events.heal.by(SELECTED_PLAYER).spell(ECHO_HEALS), this.handleEchoHeal);
+  }
+
+  handleEchoHeal(event: HealEvent) {
+    const targetID = event.targetID;
+    const spellID = event.ability.guid;
+    if (event.tick) {
+      if (!this.hotTracker.hots[targetID] || !this.hotTracker.hots[targetID][spellID]) {
+        return;
+      }
+      const hot = this.hotTracker.hots[targetID][spellID];
+      if (!this.hotTracker.fromEchoHardcast(hot) && !this.hotTracker.fromEchoTA(hot)) {
+        return;
+      }
+    } else if (!isFromTAEcho(event) && !isFromHardcastEcho(event)) {
+      return;
+    }
+    const mapRef = this.isFromTaEcho(event) ? this.taEchoHealingBySpell : this.echoHealingBySpell;
+    mapRef.set(
+      event.ability.guid,
+      mapRef.get(event.ability.guid)! + (event.amount || 0) + (event.absorbed || 0),
+    );
+  }
+
+  isFromTaEcho(event: HealEvent) {
+    const targetID = event.targetID;
+    const spellID = event.ability.guid;
+    if (event.tick) {
+      if (!this.hotTracker.hots[targetID] || !this.hotTracker.hots[targetID][spellID]) {
+        return;
+      }
+      const hot = this.hotTracker.hots[targetID][spellID];
+      return this.hotTracker.fromEchoTA(hot);
+    }
+    return isFromTAEcho(event);
+  }
+
+  get totalTaEchoHealing() {
+    let result = 0;
+    for (const amount of this.taEchoHealingBySpell.values()) {
+      result += amount;
+    }
+    return result;
+  }
+
+  get totalHardcastEchoHealing() {
+    let result = 0;
+    for (const amount of this.echoHealingBySpell.values()) {
+      result += amount;
+    }
+    return result;
+  }
+
+  totalEchoHealingForSpell(spellId: number) {
+    return (
+      (this.echoHealingBySpell.get(spellId) || 0) + (this.taEchoHealingBySpell.get(spellId) || 0)
+    );
+  }
+
+  renderDonutChart() {
```

Comment:
Reposting this here since I posted it in a resolved conversation originally: 
__________________________________________________________________________________
Thinking about this some more, if we are going to attribute fluttering seedlings from echoed emerald blossoms for this analysis, we would need to also add secondary healing gains from other talents that synergize with echo, such as;

- Additional Golden Hour healing from echoed Reversions
- Additional Lifebind healing from echoed Verdant Embraces
- 
Since that makes this even more unwieldy, the alternative I can think of is to move those contributions into their own talent analysis and specify the source of the healing (echo or hardcast) in each. Or for the sake of the donut chart combine the secondary gains into the original casted source (Reversion, Verdant Embrace, Emerald Blossom, etc)

Reply:
Yea I'll move it out, will make it much more manageable.